### DISCUSSION TITLE
daisyUI 4 upgrade guide

### DISCUSSION THREAD

**Discussion by saadeghi** [saadeghi/daisyui#2507] [General]
daisyUI 4.0 is here 🎉  
Thank you for using daisyUI and thanks for reporting issues.

If you're upgrading from daisyUI 3 to daisyUI 4, here you can see all the changes:
https://daisyui.com/docs/changelog/

![daisyUI 4 update guide](https://raw.githubusercontent.com/saadeghi/files/main/daisyui4/daisyui-41200.jpg "daisyUI 4")

In this page you can see the things you need to change:

## Themes and Colors
The biggest breaking change (as you can see in the changelog) is using OKLCH colors internally instead of HSL.  
That means if you're using the value of daisyUI color variables in your project (like `--p` for primary color) the value of those CSS variables are no longer `H S L`. They're `L C H` now and you need to use them in `oklch()` CSS color function.  
daisyUI themes are no longer supposed in old browsers ([see OKLCH compatibility chart](https://caniuse.com/?search=oklch)) and old browsers will only show a fallback default theme.

<table>
<tr>
<th> daisyUI 3 </th>
<th> daisyUI 4 </th>
</tr>
<tr>
<td>
CSS variable was HSL values

```css
--p: 251.47 100% 56%;
```
Using `hsl()` function
```css
color: hsl(var(--p))
```

</td>
<td>
CSS variable is OKLCH values

```css
--p: .4912 .3096 275.75;
```
Using `oklch()` function
```css
color: oklch(var(--p))
```

</td>
</tr>
</table>

## All `*-focus` colors are removed

You can use a `color-mix()` arbitrary class to have more control on the the colors:  
```
bg-[color-mix(in_oklab,oklch(var(--p)),black_7%)]
                   │              │       │
                   │              │       │
                   │              │       │
                   │              │       ╰── black with 7% opacity = 7% darker
                   │              │
                   │              ╰── primary color
                   │
                   ╰── OKLAB gives a better result
```
For example, here we mix each color with black to create the same colors from daisyUI 3:
<table>
<tr>
<th> daisyUI 3 </th>
<th> daisyUI 4 </th>
</tr>
<tr>
<td>

```html
<div class="bg-primary-focus">
<div class="bg-secondary-focus">
<div class="bg-accent-focus">
<div class="bg-neutral-focus">
```

</td>
<td>

```html
<div class="bg-[color-mix(in_oklab,oklch(var(--p)),black_7%)]">
<div class="bg-[color-mix(in_oklab,oklch(var(--s)),black_7%)]">
<div class="bg-[color-mix(in_oklab,oklch(var(--a)),black_7%)]">
<div class="bg-[color-mix(in_oklab,oklch(var(--n)),black_7%)]">
```

</td>
</tr>
</table>

## Extending themes

There's a change in property names in `daisyui/src/theming/themes.js` file.  
If you're customizing built-in daisyUI themes, change it:

<table>
<tr>
<th> daisyUI 3 </th>
<th> daisyUI 4 </th>
</tr>
<tr>
<td>

```js
daisyui: {
  themes: [
    {
      light: {
        ...require("daisyui/src/theming/themes")["[data-theme=light]"],
        primary: "blue",
        secondary: "teal",
      },
    },
  ],
},
```

</td>
<td>

```js
daisyui: {
  themes: [
    {
      light: {
        ...require("daisyui/src/theming/themes")["light"],
        primary: "blue",
        secondary: "teal",
      },
    },
  ],
},
```

</td>
</tr>
</table>

## RTL
All components now support LTR/RTL without any config or plugin.  
Just add `dir=rtl` to `<html>` tag and everything will be RTL on runtime.

<table>
<tr>
<th> daisyUI 3 </th>
<th> daisyUI 4 </th>
</tr>
<tr>
<td>

```js
module.exports = {
  //...
  plugins: [
    require("daisyui"),
    require("tailwindcss-flip"),
  ],
  daisyui: {
    rtl: true,
  },
}
```

</td>
<td>

```js
module.exports = {
  //...
  plugins: [
    require("daisyui"),

  ],
  daisyui: {

  },
}
```

</td>
</tr>
</table>

## Tab
 - `tabs` class is now using grid, instead of flex.
 - Lifted tab now uses a single pseudo element (`:before`) instead of two (`:before` and `after`)
 - these modifier classes for `tab` are removed:
    - `tab-lifted`
    - `tab-bordered`
    - `tab-lg`
    - `tab-md`
    - `tab-sm`
    - `tab-xs`
  - Use the following modifier classes for the parent `tabs` class instead:
    - `tabs-lifted`
    - `tabs-bordered`
    - `tabs-lg`
    - `tabs-md`
    - `tabs-sm`
    - `tabs-xs`

<table>
<tr>
<th> daisyUI 3 </th>
<th> daisyUI 4 </th>
</tr>
<tr>
<td>

```html
<div class="tabs">
  <a class="tab tab-lg tab-lifted">Tab 1</a>
  <a class="tab tab-lg tab-lifted tab-active">Tab 2</a>
  <a class="tab tab-lg tab-lifted">Tab 3</a>
  <a class="tab tab-lg tab-lifted">Tab 4</a>
</div>
```

</td>
<td>

```html
<div class="tabs tabs-lg tabs-lifted">
  <a class="tab">Tab 1</a>
  <a class="tab tab-active">Tab 2</a>
  <a class="tab">Tab 3</a>
  <a class="tab">Tab 4</a>
</div>
```

</td>
</tr>
</table>

## button
buttons are not uppercase by default anymore. You can use `uppercase` if you want to make them uppercase.

---
---

This guide will be updated if there are any missing details.  
Fill free to open issues and open PRs 💚  

--- 
PS: I'm having a lung surgery today. I will continue to fix issues as soon as I can. Thank you for your patience 🙏

---

Update: Surgery went well. Thank you all for the warm messages 💚

**Comment by vegeta897** [comment]
First of all, I hope your surgery goes well!

Thank you for making daisyUI more awesome.

I'm a little confused about custom theme support in old browsers. You say that custom themes are not supported by old browsers, and in the changelog you say "For old browsers, daisyUI provides HEX fallback colors only for the default dark/light themes". Does that mean I can define a custom theme with HEX colors and it will work in old browsers, or no? If I can, how do I define a custom theme that uses OKLCH as well as providing HEX fallbacks?

**Reply by saadeghi** [reply]
Thanks.  
I can't provide an example code right now, but here are the fallback CSS variables:  
https://github.com/saadeghi/daisyui/blob/master/src/base/colors.css

You can use a similar code to make a fallback for a custom theme

**Reply by vegeta897** [reply]
Ah I see, so there's a new set of fallback color variables. Good stuff, thanks!

**Reply by gandlafbtc** [reply]
FWIW, around 10% of active browsers still don't support OKLCH. I think it would make sense to include the theme fallback to the migration guide, to ensure that the theme looks ok on these 10% too.
 
https://caniuse.com/mdn-css_types_color_oklch

**Comment by NikarashiHatsu** [comment]
Hope you could have a godspeed recovery, thanks a lot for your efforts!

**Comment by rikgirbes** [comment]
I wish you a speedy recovery ! Hope everything goes well!

**Comment by JacobCoffee** [comment]
Thanks Pouya! Hope the surgery and recovery goes well.

**Comment by mdenjamulislam** [comment]
Thank you for making daisyUI more awesome.

**Comment by rajasegar** [comment]
Just wrote a couple of codemods to address the changes and pushed to npm
https://github.com/rajasegar/daisyui-v3-to-v4-codemods
Tested with a simple react project using Vite, codemods were working fine

**Reply by saadeghi** [reply]
Wow. Thank you 🫡

**Comment by stefanobartoletti** [comment]
Thank you for sharing this and for the v.4 update!

I suggest you to include this guide in the website docs, maybe in a page called "migration to 4" or similar, as it would probably be the first place where people go to check the docs.

I was almost opening an issue myself with a request for clarification about the migration, before checking the discussions here :)

**Comment by Metehan-Altuntekin** [comment]
I have found a difference between how the gradients are rendered. New one seems more accurate.

I am using the light theme with these classes:
`bg-gradient-to-tr from-accent from-25% via-secondary via-70% to-primary`

DaisyUI 3:
![image](https://github.com/saadeghi/daisyui/assets/86600423/bed1d166-6236-4e6a-bbf0-32919f562fdb)


DaisyUI 4:
![image](https://github.com/saadeghi/daisyui/assets/86600423/329cfaab-23d9-4cc8-9d88-de3fbfb72f43)

(Geçmiş olsun)

**Reply by seiry** [reply]
yeah, for `oklch` has a different way in gradient.

https://evilmartians.com/chronicles/oklch-in-css-why-quit-rgb-hsl#:~:text=you%E2%80%99ll%20end%20up%20with%20a%20very%20different%20gradient%20for%20the%20same%20starting%20and%20ending%20colors.

**Comment by wuarmin** [comment]
Hey @saadeghi,

can i migrate the colors during the upgrade as follows:
```css
hsl(var(--b2) / 0.2)
```
to
```css
oklch(var(--b2) / 0.2)
```
or do I have to use `color-mix` here as well?
I still don't understand when I should use color-mix to make colors darker or lighter.
Thanks
